### PR TITLE
[Builtins] Add 'MakeBuiltinMeaning' and type errors doctests

### PR DESCRIPTION
**Pull request by effectfully**
This polishes the `PlutusCore.Builtin.Meaning` module and adds another tool to `PlutusCore.Builtin.Debug` allowing us to debug the whole `BuiltinMeaning` inference pipeline.

### CODE REVIEW

**Comment by effectfully** [comment]
Sorry, don't review yet, I need to fix something.

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs [hunk 1]
```diff
@@ -236,37 +227,47 @@ type family ThrowOnBothEmpty binds args isBuiltin a where
...
-            'Text "To fix this error introduce a dummy ‘()’ argument"
+            'Text "To fix this error add a dummy ‘()’ argument"
             )
     ThrowOnBothEmpty _ _ _ _ = ()
 
--- See Note [Automatic derivation of type schemes]
--- | Construct the meaning for a built-in function by automatically deriving its
--- 'TypeScheme', given
+-- | A function turned into a type class with exactly one fully general instance.
+-- We can't package up the constraints of 'makeBuiltinMeaning' (see the instance) into a type or
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:240]: I was expecting to therefore see you replace some places that have a big load of constraints with just `MakeBuiltinMeaning`, but I can only see on in the debug module?

**Review by michaelpj — APPROVED** [review]

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs (hunk 1 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:240]: Yep, I just wanted it for `Debug`. It was constantly getting out of sync with the main thing, I got annoyed, tried to create a type/class synonym, failed, created a function-as-a-class. I can revert that if you believe it's not worth the added complexity, but it'll make me sad and what's the point of being right if your developer cries?

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs [hunk 2]
```diff
@@ -21,31 +23,60 @@ import PlutusCore.Name as Export
...
-elaborateDebug
-    :: forall a j. ElaborateFromTo 0 j (Term TyName Name DefaultUni DefaultFun ()) a
+--     • In the expression: makeBuiltinMeaningDebug $ Just ()
+-- >>> :t makeBuiltinMeaningDebug fst
+-- <interactive>:1:1: error:
+--     • Plutus type variables can't directly appear inside built-in types
+--       Are you trying to define a polymorphic built-in function over a polymorphic type?
+--       In that case you need to wrap all polymorphic built-in types having type variables
+--         in them with either ‘SomeConstant’ or ‘Opaque’ depending on whether its the type
+--         of an argument or the type of the result, respectively
```
Comment (zliu41) [inline plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs:71]: This sounds like the distinction between `SomeConstant` and `Opaque` is that one is for argument and the other is for result, but that's not the case, right?

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs [hunk 3]
```diff
@@ -21,31 +23,60 @@ import PlutusCore.Name as Export
...
---     • Built-in functions are not allowed to have constraints
---       To fix this error instantiate all constrained type variables
---     • In the second argument of ‘($)’, namely ‘0 + 42’
---       In the expression: elaborateDebug $ 0 + 42
-elaborateDebug
-    :: forall a j. ElaborateFromTo 0 j (Term TyName Name DefaultUni DefaultFun ()) a
+--     • In the expression: makeBuiltinMeaningDebug $ Just ()
+-- >>> :t makeBuiltinMeaningDebug fst
+-- <interactive>:1:1: error:
+--     • Plutus type variables can't directly appear inside built-in types
```
Comment (zliu41) [inline plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs:67]: This is confusing, because it is possible to call `makeBuiltinMeaning $ \a b -> a`, but not `makeBuiltinMeaning $ \(a, b) -> a`. Could you rephrase it to explain the difference between the two?

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs:67]: It's a type error, it's not meant to teach the user how and why things are aligned, we have proper documentation on that in the `PlutusCore.Default.Builtin` module. It is indeed unfortunate that we support `\a b -> a` and don't support `\(a, b) -> a`, but there are good reasons for that and I can't fit them into a type error, we have loads of docs with such explanations.

Even distinguishing between `a -> b -> a` and `(a, b) -> a` would require me to define what a built-in type is, talk about how `(Integer, Bool)` is different from `(a, b)` (since `(Integer, Bool) -> Integer` is perfectly fine) etc. It's just not feasible within a type error.

Think of our custom type errors as reminders. They are meant to remind you of something that you're already supposed to be familiar with and also give you quick advice on how to proceed.

Or do you maybe have a suggestion on how to improve the message?

Comment (zliu41) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs:67]: Fair enough. I don't actually know why one is allowed and the other isn't (I should revisit the docs) so I don't have a suggestion. But I was wondering if there's a good one sentence summary like "Can't call `makeBuiltinMeaning` on `\(a, b) -> a` because ...", like GHC's "because type variable `a` is untouchable". It doesn't explain what "untouchable" means (there's a paper for that) but one can then go and search for it.

But adding a builtin function is not something we frequently do so this is all fine.

**Review by zliu41 — APPROVED** [review]

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs (hunk 3 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs:67]: > "Can't call `makeBuiltinMeaning` on `\(a, b) -> a` because ...", like GHC's "because type variable `a` is untouchable".

Problem is, we piggyback on existing machinery and this is where we throw this error:

```haskell
instance TypeError NoStandalonePolymorphicDataErrMsg => uni `Contains` TyVarRep where
    knownUni = underTypeError
```

We don't have any information as to what exactly went wrong, we only know that this path is accessed when some built-in type got applied to a type variable, so we can't be specific. Maybe we can find another way of throwing the error, though.

I think "type variables can't appear inside built-in types" is a good one sentence summary, but I like the untouchable idea. But then should I add such a keyword to every other error message?

Gonna summon @michaelpj.

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs:67]: Anyways. this seems like a separate concern, so I'm gonna press "merge".

Comment (zliu41) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs:67]: Ah so "can't appear inside built-in types" in this case means "can't appear inside `(,)`". Ok, that makes sense, but I think a more explicit way of saying it would be "can't be an argument of a built-in type constructor", or "must only be used as the entire type of an argument".

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs:67]: Great suggestion, thank you. [Tweaked](https://github.com/input-output-hk/plutus/blob/177e942e4650314b52665e288bbe7538305c4a8b/plutus-core/plutus-core/src/PlutusCore/Builtin/Polymorphism.hs#L200-L205) the wording.
